File: client/src/main/kotlin/com/squareup/cash/backfila/client/RealBackfilaClient.java
package com.squareup.cash.backfila.client;

import com.squareup.moshi.Moshi;
import com.squareup.protos.backfila.service.ConfigureServiceRequest;
import com.squareup.protos.backfila.service.ConfigureServiceResponse;
import com.squareup.protos.backfila.service.ServiceType;
import com.squareup.wire.WireJsonAdapterFactory;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.Collections;
import okhttp3.OkHttpClient;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.moshi.MoshiConverterFactory;

import static com.google.common.base.Preconditions.checkState;

public class RealBackfilaClient implements BackfilaClient {
  private final BackfilaApi backfilaApi;

  public RealBackfilaClient(String host, OkHttpClient okHttpClient) {
    Moshi moshi = new Moshi.Builder()
        .add(new WireJsonAdapterFactory())
        .build();
    Retrofit retrofit = new Retrofit.Builder()
        .baseUrl(host)
        .addConverterFactory(MoshiConverterFactory.create(moshi))
        .client(okHttpClient)
        .build();
    this.backfilaApi = retrofit.create(BackfilaApi.class);
  }

  @Override public void configureService() {
    ConfigureServiceRequest request = new ConfigureServiceRequest.Builder()
        .service_type(ServiceType.CLOUD)
        .backfills(Collections.emptyList())
        .build();

      try {
        Response<ConfigureServiceResponse> response =
            backfilaApi.configureService(request).execute();
        checkState(response.isSuccessful());
      } catch (IOException e) {
        throw new UncheckedIOException(e);
      }
  }
}

File: client/src/main/kotlin/app/cash/backfila/client/RealBackfilaClient.kt
package app.cash.backfila.client

import app.cash.backfila.client.Connectors.HTTP
import app.cash.backfila.protos.service.ConfigureServiceRequest
import com.google.common.base.Preconditions.checkState
import com.squareup.moshi.Moshi
import com.squareup.wire.WireJsonAdapterFactory
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory
import java.io.IOException
import java.io.UncheckedIOException

class RealBackfilaClient(host: String, okHttpClient: OkHttpClient) : BackfilaClient {
  private val backfilaApi: BackfilaApi

  init {
    val moshi = Moshi.Builder()
        .add(WireJsonAdapterFactory())
        .build()
    val retrofit = Retrofit.Builder()
        .baseUrl(host)
        .addConverterFactory(MoshiConverterFactory.create(moshi))
        .client(okHttpClient)
        .build()
    this.backfilaApi = retrofit.create(BackfilaApi::class.java)
  }

  override fun configureService() {
    val request = ConfigureServiceRequest.Builder()
        .connector_type(HTTP)
        .backfills(emptyList())
        .build()

    try {
      val response = backfilaApi.configureService(request).execute()
      checkState(response.isSuccessful)
    } catch (e: IOException) {
      throw UncheckedIOException(e)
    }
  }
}
